### PR TITLE
Add feature: Continuous search tools (search previous, search next)

### PR DESCRIPTION
**Pull request by Steve-Mcl**
- [x] New feature (non-breaking change which adds functionality)

## Proposed changes
* Adds new search toolbar to workspace footer
  * search button is always visible and invokes the search box
  * `x of y` counter,  `prev`, `next`, `close` appear when an active search is in progress
  * pressing `prev` and `next` buttons calls the existing `reveal` function
  * pressing `close` button clears the active search and hides the footer search tools (except search button) 

### implementation notes
* Adds 2 new actions with default shortcuts...
  * `core:search-previous` : <kbd>shift</kbd>-<kbd>f</kbd>  (see "for review" below)
  * `core:search-next` : <kbd>f</kbd>  (see "for review" below)
* `RED.statusBar.addWidget` - add property `id` to the widget so it can be accessed via query selector.
  * necessary to access the search toolbar

### For review
* Added i18n entries alongside zoom controls - however - would this now make sense to be moved to `workspace.footer.[widget-id].prop: "text"` ?
* Default shortcuts are currently <kbd>f</kbd> for search next and <kbd>shift</kbd>-<kbd>f</kbd> for search previous
  * these are scoped to workspace only
  * from a UX POV, I really like the reduced context switching keyboard shortcuts offer (e.g. if I search by keyboard, I can stay on keyboard)
  * defaults are open to debate :)
* ~~Should the reveal flashing be cancelled if moving quickly through `next` / `prev`?~~
  * ~~currently, several nodes can be seen to flash when quickly searching `next` / `prev`~~
  * DONE  94e8fce
* ~~consider flashing tabs on being revealed~~
  * DONE  94e8fce
* ~~consider flashing config nodes on being revealed~~
  * DONE  94e8fce
   

## Checklist
- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
> For future reference, I find it's helpful to explicitly list any new actions added by a PR - makes writing up the docs later easier ;)

Noted :) 


> Should clicking on the magnifying glass when there is an active search open the search dialog with the 'active' search term pre-filled?

I hadnt considered that.  
I had considered using the selected node to pre-fill the search (as as it is a common pattern) but couldn't decide what to put in (name, type etc) or what to do when many were selected.  So instead, I thought clicking the magnifying glass would be akin to "new search"
Now you have thrown that spanner in the works - I feel it is equally sensible 🤔 
I'm easy & it wouldn't be difficult to implement your idea - let me know which way to go.



> Another possible tweak would be to show the search term in the footer as well...
> 
> ```
> 🔍  "potato" 1 of 27  ⬅️ ➡️ ✖️ 
> ```

My main thoughts on keeping this small as possible was for small screen devices (though I could trim with an ellipsis if > 16? chars)

Let me know if that is a good solution and I'll get this done

File: packages/node_modules/@node-red/editor-client/src/js/ui/search.js [hunk 1]
```diff
@@ -501,6 +554,8 @@ RED.search = (function() {
 
     function init() {
         RED.actions.add("core:search",show);
+        RED.actions.add("core:search-prev",revealPrev);
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply packages/node_modules/@node-red/editor-client/src/js/ui/search.js:557]: Will update to `core:search-previous`

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply packages/node_modules/@node-red/editor-client/src/js/ui/search.js:557]: Done

**Comment by knolleary** [comment]
Incorrect numbering:

image

Not 100% sure, but I think the reproduction is:

1. Search for something to get some results
2. Click 'next' a few times so you aren't on result 1
3. Search for something else - but do *not* click or press enter. Instead, click outside the search dialog to close it
4. The search bar is updated, but as we didn't select a result, the old 'current' result index is still displayed


Click outside of the search window is the same as pressing escape - it cancels the search (and the search term is *not* added to the search history). Currently this is updating the search widget. I do not think we should be doing that for a cancelled search. Cancelling the search should clear the search widget.

**Comment by Steve-Mcl** [comment]
@knolleary fingers crossed this is now ready to merge.

**Comment by knolleary** [comment]
All looks good.

My only final comment/question is... should hitting escape (closing the search dialog) go even further and clear the active search as well?

**Comment by Steve-Mcl** [comment]
> My only final comment/question is... should hitting escape (closing the search dialog) go even further and clear the active search as well?

Yes, definitely, but I couldn't think of a sensible way to capture the esc.  e.g. if the user doesnt cancel the search but then opens an editor etc.  

Thoughts?

**Comment by knolleary** [comment]
Here is where the event handlers are added for escape and click-outside to cause the search box to hide: 

https://github.com/node-red/node-red/blob/ac97e8c6136da712a53f6c024a3f7d78d6ba3ec7/packages/node_modules/%40node-red/editor-client/src/js/ui/search.js#L512-L517

They all call `hide()` directly. One approach would be to modify `hide()` to take an argument to indicate if it is a 'cancel' type hide or not. Then update the registered event handlers to call `hide` with that flag set.

Then inside `hide` you can decide whether to clear the widget or not based on that flag.

**Comment by Steve-Mcl** [comment]
Hi Nick.  Not certain that would work as expected.

consider the following chain of events...

1. user searches for "PI"
2. hits go/enter (search box and shade elements are hidden)
3. hits esc - not of the shade elements have focus and `red-ui-search` is already hidden - esc key event does not occur

feel free to correct me.

I am trying a slightly different approach. Will explain in a bit (if it works 😄 )

**Comment by knolleary** [comment]
I am not thinking about hitting escape when the search dialog is already closed. I'm specifically thinking about when the user hits escape whilst the search dialog is open - which closes the search dialog without taking any further action.

**Comment by Steve-Mcl** [comment]
OK, so i added a flag to `hide` but to "KEEP" the toolbar visible (not to close it) - this means only one call needs modifying.

Additionally, I added an `esc` keyboard shortcut (scoped to the search toolbar div) to close it when focused.  to aid this, I auto focus the NEXT or PREV button depending on search direction (so that something in the toolbar has focus) 

This feels much more natural now e.g...

1. search for "pi"
2. hit enter - search dialog is closed
3. press `F` (next)
4. press `F` (next)
5.  press `ESC` (closes toolbar)

Tested this also works for mouse actions.


Commit incoming

🤞

**Pull request merged**